### ISSUE TITLE
ping parser not working with 100% packet loss

### ISSUE THREAD

**Issue by banananananananana** [kellyjonbrazil/jc#77]
First of all, thank you for doing this. I'm doing a little project where your tool is just perfect.

With that being said, the ping parser has a problem when;
1. There is a 100% packet loss.
2. The last packet times out

With 5 packets, the three in the middle times out (i have a device in my lab i disconnected shortly) it all looks good:

> pi@raspberrypi:~ $ ping -D -O -c 5 192.168.120.164 | jc --ping | jq .
{
  "destination_ip": "192.168.120.164",
  "data_bytes": 56,
  "pattern": null,
  "destination": "192.168.120.164",
  "packets_transmitted": 5,
  "packets_received": 2,
  "packet_loss_percent": 60,
  "duplicates": 0,
  "time_ms": 149,
  "round_trip_ms_min": 1.033,
  "round_trip_ms_avg": 1.083,
  "round_trip_ms_max": 1.133,
  "round_trip_ms_stddev": 0.05,
  "responses": [
    {
      "type": "reply",
      "timestamp": 1596534506.558155,
      "bytes": 64,
      "response_ip": "192.168.120.164",
      "icmp_seq": 1,
      "ttl": 64,
      "time_ms": 1.13,
      "duplicate": false
    },
    {
      "type": "timeout",
      "timestamp": 1596534508.622236,
      "icmp_seq": 2
    },
    {
      "type": "timeout",
      "timestamp": 1596534509.662235,
      "icmp_seq": 3
    },
    {
      "type": "timeout",
      "timestamp": 1596534510.702215,
      "icmp_seq": 4
    },
    {
      "type": "reply",
      "timestamp": 1596534510.703367,
      "bytes": 64,
      "response_ip": "192.168.120.164",
      "icmp_seq": 5,
      "ttl": 64,
      "time_ms": 1.03,
      "duplicate": false
    }
  ]
}

with 5 packets, all of them times out - only 4 is parsed.

> ping -D -O -c 5 192.168.120.164 | jc --ping | jq .
{
  "destination_ip": "192.168.120.164",
  "data_bytes": 56,
  "pattern": null,
  "destination": "192.168.120.164",
  "packets_transmitted": 5,
  "packets_received": 0,
  "packet_loss_percent": 100,
  "duplicates": 0,
  "time_ms": 177,
  "responses": [
    {
      "type": "timeout",
      "timestamp": 1596534621.582227,
      "icmp_seq": 1
    },
    {
      "type": "timeout",
      "timestamp": 1596534622.622236,
      "icmp_seq": 2
    },
    {
      "type": "timeout",
      "timestamp": 1596534623.662213,
      "icmp_seq": 3
    },
    {
      "type": "timeout",
      "timestamp": 1596534624.702212,
      "icmp_seq": 4
    }
  ]
}

With 1 packet, it fails completely

>  ping -D -O -c 1 192.168.120.164 | jc --ping | jq .
jc:  Error - ping parser could not parse the input data. Did you use the correct parser?
             For details use the -d or -dd option.

I think it fails to parse the last entry if it's a timeout...

Is this something you could look into? :)

**Comment by kellyjonbrazil** [comment]
Hi there! Thank you for reporting this - I'll take a look. Could you send me the original ping output that is causing the issue? I can also try to replicate it on my end, but sometimes it's hard to get the timeouts in the right place. :)

I noticed in my initial testing that the ping command itself didn't seem to report back the last timeout, so I'm not sure if that is what is happening here. In any case, the parser should not be throwing an error like it does if no replies or timeouts are returned.  Should be an easy fix.

Also, what OS/linux distro is this on?

Thanks!

**Comment by kellyjonbrazil** [comment]
So far I'm not able to reproduce this on CentOS 7. The number of timeouts seems to be correct, since on CentOS `ping` doesn't seem to print the final timeout:
```
$ ping -D -O -c 5 192.168.120.164
PING 192.168.120.164 (192.168.120.164) 56(84) bytes of data.
[1596559421.246489] no answer yet for icmp_seq=1
[1596559422.247071] no answer yet for icmp_seq=2
[1596559423.247715] no answer yet for icmp_seq=3
[1596559424.248362] no answer yet for icmp_seq=4

--- 192.168.120.164 ping statistics ---
5 packets transmitted, 0 received, 100% packet loss, time 4002ms
```

I tried with a single ping packet and I got the expected result:
```
$ ping -D -O -c 1 192.168.120.164 | jc --ping -p
{
  "destination_ip": "192.168.120.164",
  "data_bytes": 56,
  "pattern": null,
  "destination": "192.168.120.164",
  "packets_transmitted": 1,
  "packets_received": 0,
  "packet_loss_percent": 100.0,
  "duplicates": 0,
  "time_ms": 0.0,
  "responses": []
}
```
If you could send the raw `ping` output that is failing I can probably fix it pretty quickly.

Thanks!

**Comment by banananananananana** [comment]
I just did some more testing, and I think it's a problem with the ping command.

sending 3 pings to random host on the internet that responds to icmp - i get 3 replies/lines of output
    ping -D -O -c 3 dr.dk
    PING dr.dk (159.20.6.38) 56(84) bytes of data.
    [1596561923.380531] 64 bytes from 159.20.6.38 (159.20.6.38): icmp_seq=1 ttl=250 time=5.94 ms
    [1596561924.376267] 64 bytes from 159.20.6.38 (159.20.6.38): icmp_seq=2 ttl=250 time=7.40 ms
    [1596561925.376247] 64 bytes from 159.20.6.38 (159.20.6.38): icmp_seq=3 ttl=250 time=5.58 ms

sending 3 pings to random host on the internet that blocks icmp - i get 2 timeouts/lines if output... where is icmp_seq=3 ?
    ping -D -O -c 3 jubii.dk
    PING jubii.dk (194.88.217.54) 56(84) bytes of data.
    [1596562011.254581] no answer yet for icmp_seq=1
    [1596562012.294571] no answer yet for icmp_seq=2

I'm testing on ubuntu 20.04 lts...

**Comment by kellyjonbrazil** [comment]
I pulled a couple of ping examples from my retropie and it looks like it puts an extra line break at the end sometimes. I added a quick fix to allow for that in `jc` version 1.13.3, now available on PIP. Could you try that version and see if it fixes the issue?

Thanks!

**Comment by banananananananana** [comment]
Hi

I just pulled a new version "1.13.4" and the timeout which broke the parsing is gone. Perfect. 

I'm stilling missing the actual reply/timestamp of the packet - but that's a problem/feature of iputils ping. 

Thank you so much. :)
